mne/io/base.py

```diff
@@ -973,45 +990,72 @@ def resample(self, sfreq, npad=100, window='boxcar',
+
+        for ri in range(len(inst._raw_lengths)):
+            data_chunk = inst._data[:, offsets[ri]:offsets[ri + 1]]
             new_data.append(resample(data_chunk, sfreq, o_sfreq, npad,
                                      n_jobs=n_jobs))
             new_ntimes = new_data[ri].shape[1]
 
-            # Now deal with the stim channels. In empirical testing, it was
-            # faster to resample all channels (above) and then replace the
-            # stim channels than it was to only resample the proper subset
-            # of channels and then use np.insert() to restore the stims
-
-            # figure out which points in old data to subsample
-            # protect against out-of-bounds, which can happen (having
-            # one sample more than expected) due to padding
-            stim_inds = np.minimum(np.floor(np.arange(new_ntimes) /
-                                            ratio).astype(int),
-                                   data_chunk.shape[1] - 1)
-            for sp in stim_picks:
-                new_data[ri][sp] = data_chunk[[sp]][:, stim_inds]
-
-            self._first_samps[ri] = int(self._first_samps[ri] * ratio)
-            self._last_samps[ri] = self._first_samps[ri] + new_ntimes - 1
-            self._raw_lengths[ri] = new_ntimes
-
-        # adjust affected variables
-        self._data = np.concatenate(new_data, axis=1)
-        self.info['sfreq'] = sfreq
-        self._update_times()
+            # In empirical testing, it was faster to resample all channels
+            # (above) and then replace the stim channels than it was to only
+            # resample the proper subset of channels and then use np.insert()
+            # to restore the stims.
+            stim_resampled = _resample_stim_channels(
+                data_chunk[stim_picks], new_data[ri].shape[1],
+                data_chunk.shape[1])
+            new_data[ri][stim_picks] = stim_resampled
+
+            inst._first_samps[ri] = int(inst._first_samps[ri] * ratio)
+            inst._last_samps[ri] = inst._first_samps[ri] + new_ntimes - 1
+            inst._raw_lengths[ri] = new_ntimes
+
+        inst._data = np.concatenate(new_data, axis=1)
+        inst.info['sfreq'] = sfreq
+        inst._update_times()
+
+        if events is None:
+            # Did we loose events?
+            resampled_events = find_events(inst)
+            if len(resampled_events) != len(original_events):
+                warnings.warn(
+                    'Resampling of the stim channels caused event '
+                    'information to become unreliable. Consider finding '
+                    'events on the original data and passing the event '
+                    'matrix as a parameter.'
+                )
+
+            return inst
+        else:
+            if copy:
+                events = events.copy()
+
+            events[:, 0] = np.minimum(
+                np.round(events[:, 0] * ratio).astype(int),
+                inst._data.shape[1]
```

Comment:
the round could move you outside the bounds I think

in terms of not being just `min` it's because it's an array that's being operated on